web/src/saved_replies.ts

```diff
@@ -0,0 +1,161 @@
+import $ from "jquery";
+import assert from "minimalistic-assert";
+import type * as tippy from "tippy.js";
+
+import render_add_saved_reply_modal from "../templates/add_saved_reply_modal.hbs";
+
+import * as channel from "./channel";
+import * as compose_ui from "./compose_ui";
+import * as dialog_widget from "./dialog_widget";
+import * as dropdown_widget from "./dropdown_widget";
+import type {Option} from "./dropdown_widget";
+import {$t, $t_html} from "./i18n";
+import * as keydown_util from "./keydown_util";
+import type {StateData} from "./state_data";
+import * as util from "./util";
+
+export type SavedReplies = {
+    id: number;
+    title: string;
+    content: string;
+    rendered_content: string;
+    date_created: number;
+};
+
+const ADD_SAVED_REPLY_OPTION_ID = -1;
+let my_saved_replies: SavedReplies[] = [];
+
+export function set_saved_replies(saved_replies: SavedReplies[]): void {
+    my_saved_replies = saved_replies;
+    my_saved_replies.sort((a, b) => util.strcmp(a.title.toLowerCase(), b.title.toLowerCase()));
+}
+
+export function get_saved_replies(): SavedReplies[] {
+    return my_saved_replies;
+}
+
+export function add_saved_reply(saved_reply: SavedReplies): void {
+    my_saved_replies.push(saved_reply);
+    my_saved_replies.sort((a, b) => util.strcmp(a.title.toLowerCase(), b.title.toLowerCase()));
+}
+
+export function remove_saved_reply(saved_reply_id: number): void {
+    my_saved_replies = my_saved_replies.filter((s) => s.id !== saved_reply_id);
+}
+
+export function get_options_for_dropdown_widget(): Option[] {
+    const saved_replies = my_saved_replies.map((saved_reply) => ({
+        unique_id: saved_reply.id,
+        name: saved_reply.title,
+        description: saved_reply.rendered_content.replace(/^(?:<p>)?(.*?)(?:<\/p>)?$/, "$1"),
```

Comment:
Any specific reason why the `<p>` tags being removed from the rendered_content?

Reply:
The `<p>` is removed here to show `rendered_content` in a single line.

Reply:
The `<p>` is removed here to show `rendered_content` in a single line in the dropdown.

Reply:
Ok, that makes sense.

Reply:
Hacky HTML manipulation like this can be dangerous. I'd prefer that we use CSS to manage how the rendered content gets displayed.